### DISCUSSION TITLE
Will Mojo run in browsers?

### DISCUSSION THREAD

**Discussion by AlefAlfa** [modular/modular#71] [Ideas]
Will mojo also work with web libraries like Jango, Flask or FastAPI? Will the 500x speed also be applicable when it runs in the browser? It would be awesome to just recomiple existing flask code and have it run 500x faster.

**Comment by 0arm** [comment]
It's a compiled language so it should be possible to use it for front end frameworks like yew (written in rust). Since theoretically in the future you would be able to compile it to webassembly

**Comment by TZDak** [comment]
I don't know, but just as Mojo language is meant to be a superset of Python, and Python is in a sense a scripting language, I would think it should be possible for Modular to isolate a subset of Mojo which could be supported by browsers in the same way as Javascript is, either embedded in HTML or called from HTML, which I would suggest calling Browser Mojo, although of course the decision would not be mine. As for running all of it in a browser, I would think that should require some permissions beyond what the browser itself would be likely to have and therefore likely beyond what the browser itself could grant. Definitely worthy of some discussion though and it seems that's what this space exists for... so I would encourage anyone else reading what's in this thread so far to add some of their own thoughts to it.

**Comment by ego** [comment]
I see many advantages if Mojo can be compiled into `js/ts` and these tools are close to the core or integrated into the language.

* Web space, server side rendering for SPA, one codebase for frontend/backend, other advantages.
* ML/AI prototypes for scientists aka jupyter notebook/IPython it can be replaced easily 
* Cloud provider, runtime, serverless

Python ecosystem has some tools like that:

[PyScript](https://pyscript.net/)
[brython](https://brython.info/)
[pscript](https://github.com/flexxui/pscript)
[Beeware](https://docs.beeware.org)

but Python was announced mostly as a server-side language, so it's very hard to change that mindset.

Opposed Clojure/ClojureScript by Rich Hickey has this philosophy and the community supports it, so you can find many projects built around this combination of stacks together.

We should not lose web and JavaScript developers, their community is no less than Python.

**Comment by goldiegadde** [comment]
#78  asking for similar support.

**Comment by lattner** [comment]
Maybe we should have a canonical version of this and leave it open, WDYT @goldiegadde ?

**Comment by Neon22** [comment]
PyScript is working very well in the browser. Especially with Vue3 as the UI component.
 E.g. with Vite integration - https://github.com/FabioRosado/pyscript-vue-starter

Python projects tend to do more computational work than javascript on the client side.
If Mojo was used then tuning would be super helpful to make this very very worthwhile.
(I don't pretend to understand how to dispatch the right machine level code based on what the browser is running on.)

**Comment by ksandvik** [comment]
As Mojo will be compiled to native hardware instructions, and there are security issues with browser environments, most likely the best target is wasm some time in future.

**Comment by nmn** [comment]
WASM is getting Garbage Collection. Actually, two separate sets of memory:
1. Linear Memory that has to manually managed. This is what WASM has had for a while.
2. Manage Memory: Memory shared with JS that is also garbage collected by the same GC as JS.

Considering the fact that Mojo does Python (which is memory managed) *and* manual memory management, WASM seems like a perfect compile target.

**Comment by ksandvik** [comment]
I'm no expert on MLIR but I assume someone could write Mojo MLIR primitives mapping to wasm at some point.

**Comment by ego** [comment]
There is also another solution like source-to-source S2S compilers/transcompilers/translators:

Topics:

* https://github.com/topics/transpiler
* https://github.com/topics/source-to-source

Some examples:

* [Transpiler of Python to many other languages](https://github.com/py2many/py2many)
* [Python-to-C transpiler](https://github.com/zanellia/prometeo)
* [Python to Rust transpiler](https://github.com/konchunas/pyrs)
* [Haxe can build cross-platform applications targeting JavaScript, C++, C#, Java, JVM, Python, Lua](https://haxe.org)
* [Dart supports the web as one of its core platforms. Dart-to-JavaScript compilers. Flutter](https://dart.dev/web)
* [V JavaScript and WASM backends](https://vlang.io)
* [Nim JavaScript compilation](https://nim-lang.org/features.html)

**Reply by ksandvik** [reply]
BTW wasm also opens up the possibility of 'light-weight' startup containers (wasm code) as back-end services, something that for example Fermion has prototyped out.

**Comment by lattner** [comment]
Yeah, the Mojo stack is perfectly set up to do this.  It doesn't use garbage collection, supports very small installed binaries, etc.  It'll be great, we just need to make a big more progress :)

**Comment by lattner** [comment]
Yeah, the Mojo stack is perfectly set up to do this.  It doesn't use garbage collection, supports very small installed binaries, etc.  It'll be great, we just need to make a big more progress :)
